### PR TITLE
Feature/reactions chooser

### LINKED ISSUE #112: Reactions: Add primary reactions as message actions
From riot-web created by [jryans](https://github.com/jryans): vector-im/riot-web#9576  

In the message action bar, alongside a button that opens the emoji picker to choose any emoji, we should also offer easy access to the primary emoji types:

* 👍 / 👎 
* 🙂 / 😔

### LINKED ISSUE #127: Message actions - Alt version
This action is a previous step before implementing the initial design (https://github.com/vector-im/riotX-android/issues/113)

In this implementation, we use easy and highly customisable Android bottom sheet in order to focus on the functionality instead of design.

Screenshot 2019-05-02 at 10 00 04

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #127
Fixes #112 

reactions1
reactions2

### CODE REVIEW

File: vector/src/main/res/layout/adapter_item_action_quick_reaction.xml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        android:focusable="true"
+        app:layout_constraintBottom_toTopOf="@id/quick_react_agree_text"
+        app:layout_constraintEnd_toStartOf="@id/quick_react_2"
+        app:layout_constraintHorizontal_chainStyle="packed"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_chainStyle="packed">
+
+        <TextView
```
Comment (bmarty) [inline vector/src/main/res/layout/adapter_item_action_quick_reaction.xml]: Remove RelativeLayout parent

File: vector/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -1400,4 +1402,10 @@ Why choose Riot.im?
     <string name="autodiscover_well_known_autofill_dialog_message">Riot detected a custom server configuration for your userId domain \"%s\":\n%s</string>
     <string name="autodiscover_well_known_autofill_confirm">Use Config</string>
 
+
+    <string name="title_activity_emoji_reaction_picker">Reactions</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml]: Please add new String in file string_riotX.xml

File: vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/factory/MessageItemFactory.kt [hunk 3]
```diff
@@ -86,51 +73,83 @@ class MessageItemFactory(private val colorProvider: ColorProvider,
...
-    private fun buildAudioMessageItem(messageContent: MessageAudioContent,
+    private fun buildAudioMessageItem(eventId: String, messageContent: MessageAudioContent,
                                       informationData: MessageInformationData,
                                       callback: TimelineEventController.Callback?): MessageFileItem? {
         return MessageFileItem_()
                 .informationData(informationData)
                 .filename(messageContent.body)
                 .iconRes(R.drawable.filetype_audio)
-                .clickListener { _ -> callback?.onAudioMessageClicked(messageContent) }
+                .avatarClickListener(
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/factory/MessageItemFactory.kt]: Yes but the generated models do not inherit from the models superclass, don't know how to do it :/
https://youtu.be/H7T1v3DxgKg 🎵

File: vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/item/MessageInformationData.kt [hunk 4]
```diff
@@ -18,12 +18,17 @@ package im.vector.riotredesign.features.home.room.detail.timeline.item
...
+
+@Parcelize
 data class MessageInformationData(
         val eventId: String,
         val senderId: String,
         val sendState: SendState,
         val time: CharSequence? = null,
         val avatarUrl: String?,
         val memberName: CharSequence? = null,
+        val userId: String? = null,
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotredesign/features/home/room/detail/timeline/item/MessageInformationData.kt]: Actually it's not used so I removed it. It was supposed to be myUserId, to check if the message was sent by me. But have done it another way

File: vector/src/main/java/im/vector/riotredesign/core/utils/DebouncedClickListener.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import android.view.View
+
+
+/**
+ * Simple Debounced OnClickListener
+ * Safe to use in different views
+ */
+class DebouncedClickListener(val original: View.OnClickListener, private val minimumInterval: Long = 400) : View.OnClickListener {
+    private val lastClickMap = HashMap<View, Long>()
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotredesign/core/utils/DebouncedClickListener.kt]: yep, done.

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/util/Monarchy.kt [hunk 6]
```diff
@@ -42,6 +42,17 @@ fun <T : RealmModel> Monarchy.fetchCopied(query: (Realm) -> T?): T? {
     return fetch(query, true)
 }
 
+fun <U, T : RealmModel> Monarchy.fetchMappedCopied(query: (Realm) -> T?, map: (T, realm: Realm) -> U): U? {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/util/Monarchy.kt]: done

File: vector/src/main/java/im/vector/riotredesign/features/reactions/EmojiDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import com.squareup.moshi.JsonAdapter
+
+
+
+class EmojiDataSource(val context: Context) {
+
+    var rawData: EmojiData? = null
+
+    init {
+        context.resources.openRawResource(R.raw.emoji_picker_datasource).use { input ->
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotredesign/features/reactions/EmojiDataSource.kt]: Not yet, there is a TODO for that; The reaction picker is left as is for now, but is not finished

File: vector/src/main/res/layout/adapter_item_action_quick_reaction.xml (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/adapter_item_action_quick_reaction.xml]: Done

File: vector/src/main/res/values/strings.xml (hunk 2 above)
Comment (BillCarsonFr) [reply vector/src/main/res/values/strings.xml]: Done

**Pull request merged**